py_ecc/optimized_bls12_381/optimized_pairing.py

```diff
@@ -109,30 +108,28 @@ def cast_point_to_fq12(pt: Optimized_Point3D[FQ]) -> Optimized_Point3D[FQ12]:
 
 
 # Main miller loop
-def miller_loop(Q: Optimized_Point3D[FQ12],
-                P: Optimized_Point3D[FQ12],
+def miller_loop(Q: Optimized_Point3D[FQ2],
+                P: Optimized_Point3D[FQ],
                 final_exponentiate: bool = True) -> FQ12:
     if Q is None or P is None:
         return FQ12.one()
-    R = Q  # type: Optimized_Point3D[FQ12]
+    cast_P = cast_point_to_fq12(P)
+    twist_R = twist_Q = twist(Q)
+    R = Q  # type: Optimized_Point3D[FQ2]
     f_num, f_den = FQ12.one(), FQ12.one()
     # for i in range(log_ate_loop_count, -1, -1):
     for v in pseudo_binary_encoding[62::-1]:
-        _n, _d = linefunc(R, R, P)
+        _n, _d = linefunc(twist_R, twist_R, cast_P)
         f_num = f_num * f_num * _n
         f_den = f_den * f_den * _d
         R = double(R)
+        twist_R = twist(R)
         if v == 1:
-            _n, _d = linefunc(R, Q, P)
+            _n, _d = linefunc(twist_R, twist_Q, cast_P)
             f_num = f_num * _n
             f_den = f_den * _d
             R = add(R, Q)
-        elif v == -1:
-            nQ = neg(Q)
-            _n, _d = linefunc(R, nQ, P)
-            f_num = f_num * _n
-            f_den = f_den * _d
-            R = add(R, nQ)
+            twist_R = twist(R)
     # assert R == multiply(Q, ate_loop_count)
```

Comment:
Is it okay to delete this and the following comments?

Reply:
I suppose so?

Reply:
Might be related to some of these [commented out tests](https://github.com/ethereum/py_ecc/blob/d21bbb11d4e7514956810e6ce1cd443a8e8ee817/tests/test_bn128_and_bls12_381.py#L328-L411), they're not using pytest correctly. Were they a part of the codebase back when runtime asserts *were* the test suite?